fix: Prevent query parameters from being optional when header parameters are required.

shibeshduw opened the pull request:
With this change, the file serializer in `openapi-generator` will always ensure that query parameters are not marked optional when header parameters are required by conditionally passing the correct optional flag to `serializeParamsForSignature`.

Closes SAP/cloud-sdk-js#4928.

<!-- Check List:
* Tests created/adjusted for your changes.
* PR title adheres to [conventional commit guidelines](https://www.conventionalcommits.org).
* Created a changeset `yarn changeset`
* If applicable:
  * Documented public API (TypeDoc).
  * Checked that `yarn run doc` still works.
-->

deekshas8 commented:
@shibeshduw I am not sure if I am misunderstanding, but I understood the user's issue as: incorrect code is generated when there are optional query and required header params, I thought this was something we covered in our [tests](https://github.com/SAP/cloud-sdk-js/blob/5847e0af7e04286e58a6794bc297dffe5895430d/packages/openapi-generator/src/file-serializer/operation.spec.ts#L195). 

Is this not the case?

shibeshduw commented:
Yes you're correct. But the manner in which we were [checking for optional modifier](https://github.com/SAP/cloud-sdk-js/blob/5847e0af7e04286e58a6794bc297dffe5895430d/packages/openapi-generator/src/file-serializer/operation.ts#L47) was wrong. I was suggesting that we make the order of parameters dynamic but after speaking with @marikaner , I go to know that this would be a breaking change. So now I have reverted the changes and added a minimal fix for the issue.

marikaner commented on packages/openapi-generator/src/file-serializer/operation.ts:
```diff
@@ -36,15 +36,24 @@ ${operation.operationId}: (${serializeOperationSignature(
     operation,
     'queryParameters',
-    headerParams?.includes('?')
+    allOptionalHeaders ? allOptionalQuery : false
```
```suggestion
    allOptionalHeaders && allOptionalQuery
```

marikaner reviewed (approved):
LGTM

marikaner commented:
Please still correct the title 😊
